kitsune/kpi/static/kpi/js/components/Chart.es6.js

```diff
@@ -0,0 +1,182 @@
+/* globals d3:false, $:false, _:false */
+/* jshint esnext: true */
+
+export default class Chart {
+  constructor($container, options) {
+    let defaults = {
+      chartColors: ['#EE2820', '#F25743', '#F58667', '#F9B58B', '#FDE4AF', '#E3F1B6', '#AADA9F', '#72C489', '#39AD72', '#00975C'],
+      axes: {
+        xAxis: {
+          labels: [],
+          labelOffsets: { x: 15, y: 23 }
+        },
+        yAxis: {
+          labels: [],
+          labelOffsets: { x: 5, y: 23 }
+        },
+        getPosition: (position, axis, index) => {
+          let gridSize = (position === 'y') ? this.gridSize/2 : this.gridSize;
+
+          if (position === axis[0]) {
+            return index * gridSize;
+          } else {
+            return -gridSize;
+          }
+        }
+      },
+      margin: { top: 40, right: 0, bottom: 100, left: 75 },
+      width: 860,
+      height: 430,
+      grid: { rows: 12, columns: 12 },
+      gridSize: 71,
+      buckets: 10,
+      legendElementWidth: 95,
+      data: [],
+      dom: {
+        graphContainer: $container.find('.graph').get()[0]
+      }
+    };
+
+    $.extend(true, this, defaults, options);
+
+    this.init();
+  }
+
+  // Render whatever pieces of the chart we can while waiting for data
+  preRender() {
+    // draw the container svg for the chart
+    this.dom.svg = d3.select(this.dom.graphContainer).append('svg')
+      .attr('width', this.width + this.margin.left + this.margin.right)
+      .attr('height', this.height + this.margin.top + this.margin.bottom)
+      .append('g')
+        .attr('transform', 'translate(' + this.margin.left + ',' + this.margin.top + ')');
+
+    this.dom.svg.append('g')
+      .attr('class', 'data');
+
+    this.setupAxis('xAxis');
+    this.setupLegend();
+  }
+
+  setupAxis(axis) {
+    let self = this;
+    let axisGroup = self.dom.svg.append('g')
+      .attr('class', axis);
+
+    axisGroup.selectAll()
+      .data(self.axes[axis].labels)
+        .enter().append('text')
+        .text(function(d,i) { return d; })
+        .attr('x', function(d, i) {
+          return self.axes.getPosition('x', axis, i) + self.axes[axis].labelOffsets.x;
+        })
+        .attr('y', function(d, i) {
+          return self.axes.getPosition('y', axis, i) + self.axes[axis].labelOffsets.y;
+        })
+  }
+
+  setupLegend() {
+    let self = this;
```

Comment:
So I don't need it everywhere in the methods, but when you get into the d3 stuff like setting x and y attributes, the `this` context changes to the d3 selection.